File: 2021/05/14/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if root == None or (root.left == None and root.right == None):
            return root
        if root.left == None:
            return self.flatten(root.right)
        
        if root.right == None:
            root.right = root.left
            root.left = None
            return self.flatten(root.right)
        
        r = self.flatten(root.right)
        self.flatten(root.left).right = root.right
        root.right = root.left
        root.left = None
        return r

File: 2021/05/14/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func flatten(root *TreeNode) *TreeNode {
	if root == nil || (root.Left == nil && root.Right == nil) {
		return root
	}
	if root.Left == nil {
		return flatten(root.Right)
	}
	if root.Right == nil {
		root.Right = root.Left
		root.Left = nil
		return flatten(root.Right)
	}
	r := flatten(root.Right)
	flatten(root.Left).Right = root.Right
	root.Right = root.Left
	root.Left = nil
	return r
}
